### ISSUE TITLE
Support TFP's (L-)BFGS optimiser

### ISSUE THREAD

**Issue by vdutor** [GPflow/GPflow#1604]
# Feature request: Support TFP's (L-)BFGS optimiser

TensorFlow Probability (TFP) has a native TensorFlow implementation of the (L-)BFGS optimiser. Currently, GPflow doesn't use this and relies on scipy's implementation. The problem with this, however, is that in each step of the optimisation we have to leave "graph" mode and pass in NumPy values directly to SciPy - possibly introducing a computational inefficiency.   

## Proposal

It would be nice to make TFP's optimisers available in GPflow. Here's a quick and dirty stab at it:
```python
# Copyright © 2019 Pi-Yueh Chuang <pychuang@gwu.edu>
# Copyright © 2017-2020 The GPflow Contributors. All Rights Reserved.
# Copyright © 2020 Vincent Dutordoir <vincent@secondmind.ai>
#
# Distributed under terms of the MIT license.


from typing import Sequence, Union, List

import gpflow
import numpy as np
import tensorflow as tf
import tensorflow_probability as tfp
from matplotlib import pyplot as plt


def pack_tensors(tensors: Sequence[Union[tf.Tensor, tf.Variable]]) -> tf.Tensor:
    flats = [tf.reshape(tensor, (-1,)) for tensor in tensors]
    tensors_vector = tf.concat(flats, axis=0)
    return tensors_vector


def unpack_tensors(
    to_tensors: Sequence[Union[tf.Tensor, tf.Variable]], from_vector: tf.Tensor
) -> List[tf.Tensor]:
    s = 0
    values = []
    for target_tensor in to_tensors:
        shape = tf.shape(target_tensor)
        dtype = target_tensor.dtype
        tensor_size = tf.reduce_prod(shape)
        tensor_vector = from_vector[s : s + tensor_size]
        tensor = tf.reshape(tf.cast(tensor_vector, dtype), shape)
        values.append(tensor)
        s += tensor_size
    return values


def assign_tensors(to_tensors: Sequence[tf.Variable], values: Sequence[tf.Tensor]) -> None:
    if len(to_tensors) != len(values):
        raise ValueError("to_tensors and values should have same length")
    for target, value in zip(to_tensors, values):
        target.assign(value)


def create_value_and_gradient_function(loss_closure, trainable_variables, verbose=1):
    """A factory to create a function required by tfp.optimizer.lbfgs_minimize.
    Args:
        loss_closure:
        trainable_variables:
    Returns:
        A function that has a signature of:
            loss_value, gradients = f(model_parameters).
    """
    i = tf.Variable(0)

    def assign_param_values_to_variables(x):
        values = unpack_tensors(trainable_variables, x)
        assign_tensors(trainable_variables, values)

    @tf.function
    def f_value_and_gradients(x):
        """A function that can be used by tfp.optimizer.lbfgs_minimize"""
        # update params
        assign_param_values_to_variables(x)
        # compute loss value and gradients w.r.t. trainable variables
        with tf.GradientTape(watch_accessed_variables=False) as tape:
            tape.watch(trainable_variables)
            loss_value = loss_closure()
        grads = tape.gradient(loss_value, trainable_variables)

        i.assign_add(1)
        if verbose > 0:
            tf.print("Iter:", i, "loss:", loss_value)

        # return loss and flattened gradients
        return loss_value, pack_tensors(grads)

    return f_value_and_gradients, assign_param_values_to_variables


def plot_helper(inputs, outputs, title):
    """Plot helper"""
    plt.figure()
    plt.tricontourf(inputs[:, 0], inputs[:, 1], outputs.flatten(), 100)
    plt.xlabel("x")
    plt.ylabel("y")
    plt.title(title)
    plt.colorbar()


if __name__ == "__main__":
    # use float64 by default
    tf.keras.backend.set_floatx("float64")

    # prepare training data
    x_1d = np.linspace(-1.0, 1.0, 11)
    x1, x2 = np.meshgrid(x_1d, x_1d)
    inps = np.stack((x1.flatten(), x2.flatten()), 1)
    outs = np.reshape(inps[:, 0] ** 2 + inps[:, 1] ** 2, (x_1d.size ** 2, 1))
    data = (inps, outs)

    model = gpflow.models.GPR(data, gpflow.kernels.RBF())
    loss_closure = model.training_loss_closure()

    func, assign = create_value_and_gradient_function(
        loss_closure=loss_closure, trainable_variables=model.trainable_variables,
    )
    print("Before training", loss_closure())

    # # convert initial model parameters to a 1D tf.Tensor
    init_params = pack_tensors(model.trainable_variables)

    # # train the model with L-BFGS solver
    results = tfp.optimizer.lbfgs_minimize(
        value_and_gradients_function=func, initial_position=init_params, max_iterations=5
    )

    # after training, the final optimized parameters are still in results.position
    # so we have to manually put them back to the model
    assign(results.position)
    print("After training", loss_closure())

    # # do some prediction
    pred_outs = model.predict_f(inps)[0].numpy()
    err = np.abs(pred_outs - outs)
    print("L2-error norm: {}".format(np.linalg.norm(err) / np.sqrt(11)))

    # plot figures
    plot_helper(inps, outs, "Exact solution")
    plot_helper(inps, pred_outs, "Predicted solution")
    plot_helper(inps, err, "Absolute error")
    plt.show()
```

**Comment by vdutor** [comment]
pinging @gustavocmv and @st--  as I believe this will interest you.

**Comment by gustavocmv** [comment]
Thanks for the ping, @vdutor !

I have used this code myself and it helped me a lot. In fact, I have built a class-wrapper around it to make it look like other TensorFlow optimizers, with an iteration-based `.minimize` method. I'm pasting its code below to maybe start some design discussion.
```python
class LBFGSOptimizer():
​
    def __init__(self, loss_closure, trainable_variables, steps=1):
        tf.keras.backend.set_floatx("float64")
        self.initial_position = pack_tensors(trainable_variables)
        self.results = None
        func, assign = create_value_and_gradient_function(
            loss_closure=loss_closure,
            trainable_variables=trainable_variables
        )
        self.func = func
        self.assign = assign
        self.steps = steps
​
    @property
    def epoch(self):
        if self.results is None:
            return 0
        return int(self.results.num_iterations.numpy())
​
    @property
    def loss(self):
        if self.results is None:
            return None
        return float(self.results.objective_value.numpy())
​
    def minimize(self):
        if self.results is None:
            initial_poisition = self.initial_position
        else:
            initial_poisition = None
        self.results = tfp.optimizer.lbfgs_minimize(
            value_and_gradients_function=self.func,
            initial_position=initial_poisition,
            previous_optimizer_results=self.results,
            max_iterations=tf.cast(self.epoch + self.steps, dtype=tf.int32)
        )
        self.assign(self.results.position)
```
I made it in a rush, so there are some calls to `.numpy()` and type-castings which could be removed. What do you think about it?

**Comment by st--** [comment]
Just for clarification, there **is** support for (L-)BFGS(-B) and other second-order optimizers (e.g. Conjugate Gradients) via [`gpflow.optimizers.Scipy`](https://gpflow.readthedocs.io/en/master/gpflow/optimizers/index.html#gpflow-optimizers-scipy), which connects TensorFlow's gradient computation with [`scipy.optimize.minimize`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.minimize.html) (which is pure numpy). This issue is about additional support for TensorFlow Probability's *native* (TensorFlow-only, no calling into numpy land) [(L-)BFGS optimizer](https://www.tensorflow.org/probability/api_docs/python/tfp/optimizer/lbfgs_minimize).
